misc/metrics/src/swarm.rs

```diff
@@ -156,8 +157,9 @@ impl<TBvEv, THandleErr> super::Recorder<libp2p_swarm::SwarmEvent<TBvEv, THandleE
                     })
                     .inc();
             }
-            libp2p_swarm::SwarmEvent::IncomingConnection { .. } => {
-                self.connections_incoming.inc();
+            libp2p_swarm::SwarmEvent::IncomingConnection { send_back_addr, .. } => {
+                let labels = protocol_stack(send_back_addr);
+                self.connections_incoming.get_or_create(&labels).inc();
```

Comment:
I don't think we want to create one label per protocol. Instead we want one label per entire stack, i.e. the labels here should be all concatenated.

This is probably best to happen inside the `protocol_stack` function so it only returns a `String`.

Reply:
That actually raises a risk of collision (not right now, but a potential future tag) unless a delimiter is used. The most obvious delimiter would be "/".

After we go to `&str` it should be a similar number of allocations between `Vec<&'static str>` and `String`, correct? Are you thinking comparisons and hashing is faster with String or something of that nature?

Reply:
> That actually raises a risk of collision (not right now, but a potential future tag) unless a delimiter is used. The most obvious delimiter would be "/".

Yeah, the original issue uses `/` as a delimiter: https://github.com/libp2p/rust-libp2p/issues/2758

> After we go to `&str` it should be a similar number of allocations between `Vec<&'static str>` and `String`, correct? Are you thinking comparisons and hashing is faster with String or something of that nature?

The allocations I was concerned with were the ones were we allocate a new string _for each_ protocol in the multiaddr. Constructing a new, single string will require an allocation. I don't think we can get rid of that one.

Unfortunately, [`intersperse`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.intersperse) is not yet stable, otherwise we could have done:

```rust
addr
	.iter()
	.map(|p| match p {
		Protocol::Tcp(_) => "tcp",
	 	// etc
	})
	.intersperse("/")
	.collect::<String>();
``` 

Actually, looking at the code further, there isn't really a point in doing that because `String::extend` [pushes each item individually](https://doc.rust-lang.org/src/alloc/string.rs.html#2073), meaning the `String` will re-allocate as we push more items. Iterating twice to know the length also doesn't seem to be worth it so I think we will have to do something like this:

```rust
let mut protocols = addr
	.iter()
	.map(|p| match p {
		Protocol::Tcp(_) => "tcp",
	 	// etc
	});

let mut stack = match protocols.next() {
	Some(first_proto) => String::from(first_proto),
	None => return String::new()
};

while let Some(proto) = protocols.next() {
	stack.push_str("/");
	stack.push_str(proto);
}

stack
```

Alternatively, we could depend on `itertools` until `intersperse` is stable. It is "just" the `libp2p-metrics` crate so I don't mind much. @mxinden thoughts?